### Layer Initialization and Transfer Learning

Data normalization scales input features for stability; similarly, the weights of a linear layer are also initialized to small values. This is known as ***Layer Initialization***.

> Let's create a small linear layer and check its weight range.

In [2]:
import torch.nn as nn

layer = nn.Linear(64, 128)
print(layer.weight.min(), layer.weight.max())

tensor(-0.1250, grad_fn=<MinBackward1>) tensor(0.1249, grad_fn=<MaxBackward1>)


We observe that the weights are between ***-0.125*** and ***0.125***. 

Why is it important? the output of a neuron in a linear layer is a weighted sum of inputs from the previous layer. Keeping both the input data and layer weights small ensures stable outputs, preventing extreme values that could slow training. Layers can be intialized in different ways and it remains an active area of research.

---

PyTorch provides an easy way to initialize layers weights with the ***nn.init*** module. For example, here we initialize a linear layer with an uniform distribution. As you can see, the weights values now range from 0 to 1.

In [3]:
import torch.nn as nn

layer = nn.Linear(64, 128)
nn.init.uniform_(layer.weight)

print(layer.weight.min(), layer.weight.max())

tensor(0.0002, grad_fn=<MinBackward1>) tensor(0.9999, grad_fn=<MaxBackward1>)
